### PR TITLE
Use a single Scarpe::App instance as self nearly everywhere

### LINKED ISSUE #136: Shoes DSL: self in blocks, Shoes::Widget, different APIs in different places
In Nobody Knows Shoes, page 36 (or possibly 23? some of the page numbering stuff seems wonky), he says that if you were to inspect self in any block, you'd see an App object. So the way we're using instance_eval is wrong, in the sense that it's not how Shoes does it. I'm not sure how he tracks the current widget context to build the widget tree, though of course there are ways to do it.

This isn't a purely academic issue. If self is always App, then if you set an instance variable in Shoes Classic, that instance variable is available everywhere in your program. That's not true for us. D'oh!

### PR DESCRIPTION
**Pull request by noahgibbs**
See issue #136 for initial bug report. See https://github.com/scarpe-team/scarpe/wiki/ShoesSlotsSelf.md for a description of the preferred behaviour.

This should fix #136 and #250 when complete and merged.

### Description

This PR passes the Shoes::App instance as self to relevant APIs.

However, I did *not* do this for Shape. Shape is going to require a more extensive rework. I'll file a bug for it.

### Checklist

- [x] Run tests locally
- [x] Run linter(check for linter errors)

### CODE REVIEW

File: lib/scarpe/download.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+require "net/http"
+require "openssl"
+require "nokogiri"
+
+class Scarpe::Widget
+  def download(url)
```
Comment (noahgibbs) [inline lib/scarpe/download.rb:8]: Note to me: make sure we have a test for download. Is this on the right class?

**Review by Schwad — APPROVED** [review]
Paired on this

**Comment by noahgibbs** [comment]
Notes from review:

Add test that uses ivar properly, including set inside stack and used outside it.
Make sure event handler for button-click uses the right self.

**Comment by Schwad** [comment]
Will try to rewrite some examples without globals after

**Comment by noahgibbs** [comment]
Okay. Added the new tests, will merge after CI builds.

**Pull request merged**